src/mito2/src/compaction/compactor.rs

```diff
@@ -0,0 +1,418 @@
+                            index_write_buffer_size,
+                            index_options,
+                        },
+                        &write_opts,
+                    )
+                    .await?
+                    .map(|sst_info| FileMeta {
+                        region_id,
+                        file_id,
+                        time_range: sst_info.time_range,
+                        level: output.output_level,
+                        file_size: sst_info.file_size,
+                        available_indexes: sst_info
+                            .inverted_index_available
+                            .then(|| SmallVec::from_iter([IndexType::InvertedIndex]))
+                            .unwrap_or_default(),
+                        index_file_size: sst_info.index_file_size,
+                    });
+                Ok(file_meta_opt)
+            });
+        }
+        let mut output_files = Vec::with_capacity(futs.len());
+        while !futs.is_empty() {
+            let mut task_chunk =
+                Vec::with_capacity(crate::compaction::task::MAX_PARALLEL_COMPACTION);
+            for _ in 0..crate::compaction::task::MAX_PARALLEL_COMPACTION {
+                if let Some(task) = futs.pop() {
+                    task_chunk.push(common_runtime::spawn_bg(task));
+                }
+            }
+            let metas = futures::future::try_join_all(task_chunk)
+                .await
+                .context(JoinSnafu)?
+                .into_iter()
+                .collect::<Result<Vec<_>>>()?;
+            output_files.extend(metas.into_iter().flatten());
+        }
+
+        let mut inputs: Vec<_> = compacted_inputs.into_iter().collect();
+        inputs.extend(
+            picker_output
+                .expired_ssts
+                .iter()
+                .map(|f| f.meta_ref().clone()),
+        );
+
+        Ok(MergeOutput {
+            files_to_add: output_files,
+            files_to_remove: inputs,
+            compaction_time_window: Some(picker_output.time_window_size),
+        })
+    }
+
+    async fn update_manifest(
+        &self,
+        compaction_region: &CompactionRegion,
+        merge_output: MergeOutput,
+    ) -> Result<()> {
+        if merge_output.is_empty() {
+            return Ok(());
+        }
+
+        // Write region edit to manifest.
+        let edit = RegionEdit {
+            files_to_add: merge_output.files_to_add,
+            files_to_remove: merge_output.files_to_remove,
+            compaction_time_window: merge_output
+                .compaction_time_window
+                .map(|seconds| Duration::from_secs(seconds as u64)),
+            flushed_entry_id: None,
+            flushed_sequence: None,
+        };
+
+        let action_list = RegionMetaActionList::with_action(RegionMetaAction::Edit(edit.clone()));
+        // TODO: We might leak files if we fail to update manifest. We can add a cleanup task to remove them later.
+        compaction_region
```

Comment:
If compactor is running as a standalone process, then it may not be safe to directly update manifest after compaction.

Reply:
The `compact()` is not for remote compaction(I think I should add the comment). For example:

```
...
#[async_trait]
impl Tool for Compact {
    async fn do_work(&self) -> Result<()> {
        let object_store_manager = self.create_object_store_manager();
        let compactor = DefaultCompactor {};
        let compaction_region =
            open_compaction_region(&self.request, self.data_home.as_str(), object_store_manager)
                .await
                .context(OpenCompactRegionSnafu)?;

        compactor
            .compact(compaction_region, self.request.compaction_options.clone())
            .await
            .context(CompactRegionSnafu)?;

        Ok(())
    }
}
...
```